File: quests/chapter10/python/security.py
# Reference solutions for Chapter 10: Citadel of Firewalls
# These are working implementations for testing purposes

import hashlib
import json
import base64
import re
from typing import Dict, Any, List


def hash_password(password: str, salt: str) -> str:
    """Hash a password with a salt using SHA256.

    Args:
        password: The password to hash
        salt: The salt to combine with password

    Returns:
        Hex string of the hash
    """
    combined = password + salt
    hash_obj = hashlib.sha256(combined.encode())
    return hash_obj.hexdigest()


def validate_jwt(token: str) -> Dict[str, Any]:
    """Validate JWT token structure and decode payload.

    JWT format: header.payload.signature

    Args:
        token: The JWT token to validate

    Returns:
        Dict with 'valid' boolean and 'payload' if valid
    """
    parts = token.split(".")

    if len(parts) != 3:
        return {"valid": False}

    try:
        header_b64, payload_b64, signature = parts

        # Validate that parts are non-empty
        if not header_b64 or not payload_b64 or not signature:
            return {"valid": False}

        # Decode header (should be valid JSON)
        header_json = base64.urlsafe_b64decode(header_b64 + "==")
        header = json.loads(header_json)

        # Decode payload (should be valid JSON)
        payload_json = base64.urlsafe_b64decode(payload_b64 + "==")
        payload = json.loads(payload_json)

        # Basic validation passed
        return {"valid": True, "payload": payload}
    except Exception:
        return {"valid": False}


def rate_limit(requests: List[Dict[str, Any]], limit: int, window_ms: int) -> bool:
    """Check if rate limit is exceeded for requests.

    Args:
        requests: List of dicts with 'ip' and 'timestamp' keys
        limit: Maximum number of requests allowed in window
        window_ms: Time window in milliseconds

    Returns:
        True if rate limit is exceeded, False otherwise
    """
    import time

    # Calculate the earliest timestamp in the window
    now = int(time.time() * 1000)  # Current time in milliseconds
    window_start = now - window_ms

    # Count requests within the window
    request_count = sum(1 for req in requests if req.get("timestamp", 0) >= window_start)

    # Return true if limit is exceeded
    return request_count > limit


def sanitize_input(input_str: str) -> str:
    """Sanitize input to prevent XSS attacks.

    Removes script tags and HTML tags, escapes special characters.

    Args:
        input_str: The input string to sanitize

    Returns:
        Sanitized string safe from XSS attacks
    """
    # Remove script tags and their content
    sanitized = re.sub(r"<script\b[^<]*(?:(?!</script>)<[^<]*)*</script>", "", input_str, flags=re.IGNORECASE)

    # Remove other HTML tags
    sanitized = re.sub(r"<[^>]*>", "", sanitized)

    # Escape dangerous characters
    sanitized = sanitized.replace('"', "&quot;").replace("'", "&#39;")

    # Fix ampersands (must be done last to avoid double-escaping)
    sanitized = re.sub(r"&(?!(?:[a-z]+|#[0-9]+);)", "&amp;", sanitized, flags=re.IGNORECASE)

    return sanitized

File: quests/chapter10/java/Security.java
// Chapter 10: Citadel of Firewalls - Security
// Your task: Implement security fundamentals: password hashing, JWT validation, XSS prevention

import java.util.Map;
import java.util.HashMap;

public class Security {

    /**
     * Hash a password using SHA-256
     *
     * Password hashing is a one-way function that converts a password into a fixed-length
     * string of characters. This is crucial for security because:
     * - Passwords should never be stored in plain text
     * - Even if the database is compromised, attackers can't easily recover passwords
     * - SHA-256 produces a 64-character hexadecimal string
     *
     * TODO: Implement password hashing using SHA-256
     * - Use java.security.MessageDigest to create a SHA-256 hash
     * - Convert the password string to bytes
     * - Return the hash as a hexadecimal string
     *
     * Example:
     *   hashPassword("secret123") -> "fcf730b6d95236ecd3c9fc2d92d7b6b2bb061514961aec041d6c7a7192f592e4"
     *
     * @param password The plain text password to hash
     * @return The SHA-256 hash as a hexadecimal string
     */
    public static String hashPassword(String password) {
        // TODO: Implement SHA-256 password hashing
        // Hint: Use MessageDigest.getInstance("SHA-256")
        // Then convert byte array to hex string
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Verify a password against a stored hash
     *
     * This function compares a plain text password with a previously stored hash
     * to determine if they match.
     *
     * TODO: Implement password verification
     * - Hash the provided password using the same algorithm
     * - Compare the result with the stored hash
     * - Return true if they match, false otherwise
     *
     * Example:
     *   verifyPassword("secret123", "fcf730b6d95236ecd3c9fc2d92d7b6b2bb061514961aec041d6c7a7192f592e4") -> true
     *   verifyPassword("wrongpass", "fcf730b6d95236ecd3c9fc2d92d7b6b2bb061514961aec041d6c7a7192f592e4") -> false
     *
     * @param password The plain text password to verify
     * @param hash The stored hash to compare against
     * @return true if the password matches the hash, false otherwise
     */
    public static boolean verifyPassword(String password, String hash) {
        // TODO: Implement password verification
        // Hint: Hash the password and compare with the stored hash
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Sanitize user input to prevent XSS (Cross-Site Scripting) attacks
     *
     * XSS attacks occur when malicious scripts are injected into web pages.
     * Sanitization escapes dangerous HTML characters to prevent code execution.
     *
     * Characters to escape:
     * - & -> &amp;
     * - < -> &lt;
     * - > -> &gt;
     * - " -> &quot;
     * - ' -> &#39;
     *
     * TODO: Implement input sanitization
     * - Replace all dangerous characters with their HTML entity equivalents
     * - Process & first to avoid double-escaping
     *
     * Example:
     *   sanitizeInput("<script>alert('XSS')</script>") -> "&lt;script&gt;alert(&#39;XSS&#39;)&lt;/script&gt;"
     *
     * @param input The user input to sanitize
     * @return The sanitized string with HTML entities escaped
     */
    public static String sanitizeInput(String input) {
        // TODO: Implement XSS prevention through character escaping
        // Hint: Replace dangerous characters in the correct order
        // Important: Handle & first to avoid double-escaping
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Validate JWT (JSON Web Token) structure
     *
     * A JWT consists of three parts separated by dots:
     * 1. Header - Contains token type and algorithm (Base64 encoded JSON)
     * 2. Payload - Contains claims/data (Base64 encoded JSON)
     * 3. Signature - Verifies the token hasn't been tampered with
     *
     * Format: xxxxx.yyyyy.zzzzz
     *
     * TODO: Implement JWT structure validation
     * - Split the token by '.'
     * - Verify there are exactly 3 parts
     * - Verify each part is non-empty
     *
     * Note: This only validates the STRUCTURE, not the signature.
     * In production, you would also verify the cryptographic signature.
     *
     * Example:
     *   validateJWTStructure("eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxMjM0NTY3ODkwIn0.abc123") -> true
     *   validateJWTStructure("invalid.token") -> false
     *   validateJWTStructure("only.two.") -> false (empty third part)
     *
     * @param token The JWT token string to validate
     * @return true if the token has valid structure, false otherwise
     */
    public static boolean validateJWTStructure(String token) {
        // TODO: Implement JWT structure validation
        // Hint: Split by "." and check for exactly 3 non-empty parts
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Decode the payload from a JWT token
     *
     * The payload is the second part of the JWT, containing the claims.
     * It is Base64URL encoded JSON that needs to be decoded.
     *
     * TODO: Implement JWT payload decoding
     * - Validate the JWT structure first
     * - Extract the second part (payload)
     * - Decode from Base64URL
     * - Parse the JSON into a Map
     *
     * Note: JWT uses Base64URL encoding (with - and _ instead of + and /)
     * Java's Base64.getUrlDecoder() handles this automatically.
     *
     * Example JWT payload (decoded): {"sub":"1234567890","name":"John Doe","iat":1516239022}
     *
     * @param token The JWT token string
     * @return A Map containing the decoded payload claims, or null if invalid
     */
    public static Map<String, Object> decodeJWTPayload(String token) {
        // TODO: Implement JWT payload decoding
        // Hint:
        // 1. Validate structure with validateJWTStructure()
        // 2. Split and get the middle part
        // 3. Use Base64.getUrlDecoder().decode()
        // 4. Parse JSON (can use simple parsing or a JSON library)
        throw new UnsupportedOperationException("Not implemented yet");
    }
}
